hivemind/p2p/p2p_daemon.py

```diff
@@ -0,0 +1,290 @@
+import asyncio
+import copy
+from pathlib import Path
+import pickle
+import subprocess
+import typing as tp
+
+import google.protobuf
+from multiaddr import Multiaddr
+import hivemind.p2p.p2p_daemon_bindings.p2pclient as p2pclient
+from hivemind.p2p.p2p_daemon_bindings.datastructures import ID, StreamInfo
+
+from hivemind.utils.networking import find_open_port
+from hivemind.utils.logging import get_logger
+
+
+logger = get_logger(__name__)
+
+
+class P2PContext(object):
+    def __init__(self, ours_id, ours_port, handle_name):
+        self.peer_id = None
+        self.peer_addr = None
+        self.ours_id = ours_id
+        self.ours_port = ours_port
+        self.handle_name = handle_name
+
+
+class P2P(object):
+    """
+    Forks a child process and executes p2pd command with given arguments.
+    Can be used for peer to peer communication and procedure calls.
+    Sends SIGKILL to the child in destructor.
+    """
+
+    P2PD_RELATIVE_PATH = 'hivemind_cli/p2pd'
+    NUM_RETRIES = 3
+    RETRY_DELAY = 0.4
+    HEADER_LEN = 8
+    BYTEORDER = 'big'
+
+    class IncompleteRead(Exception):
+        pass
+
+    class InterruptedError(Exception):
+        pass
+
+    def __init__(self):
+        self._child = None
+        self._listen_task = None
+        self._server_stopped = asyncio.Event()
+
+    @classmethod
+    async def create(cls, *args, quic=1, tls=1, conn_manager=1, dht_client=1,
+                     nat_port_map=True, auto_nat=True, bootstrap=True,
+                     host_port: int = None, daemon_listen_port: int = None, **kwargs):
+        self = cls()
+        p2pd_path = Path(__file__).resolve().parents[1] / P2P.P2PD_RELATIVE_PATH
+        proc_args = self._make_process_args(
+            str(p2pd_path), *args,
+            quic=quic, tls=tls, connManager=conn_manager,
+            dhtClient=dht_client, natPortMap=nat_port_map,
+            autonat=auto_nat, b=bootstrap, **kwargs)
+        self._assign_daemon_ports(host_port, daemon_listen_port)
+        for try_count in range(self.NUM_RETRIES):
+            try:
+                self._initialize(proc_args)
+                await self._identify_client(P2P.RETRY_DELAY * (2 ** try_count))
+            except Exception as e:
+                logger.debug(f"Failed to initialize p2p daemon: {e}", RuntimeWarning)
```

Comment:
This will fail with `TypeError: not all arguments converted during string formatting` since the `RuntimeWarning` argument is not used.